File: swarmmanagercontroller/src/main/java/com/swarmmanager/docker/api/common/AbstractApiImpl.java
package com.swarmmanager.docker.api.common;

import com.swarmmanager.docker.api.common.client.DockerWebClient;
import com.swarmmanager.docker.api.common.parameters.FiltersParameters;
import com.swarmmanager.docker.api.common.parameters.HeaderParameters;
import com.swarmmanager.docker.api.common.parameters.QueryParameters;
import com.swarmmanager.docker.api.common.parameters.RequestBodyParameter;
import com.swarmmanager.rest.*;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractApiImpl {

    @Autowired
    protected DockerWebClient dockerWebClient;

    protected  <E> List<E> listObjects(String apiPath, String swarmId, RestResponseType<List<E>> restResponseType, FiltersParameters filtersParameters) {
        if (filtersParameters != null) {
            RestParameters restParameters = new RestParameters(dockerWebClient.getBaseResource(swarmId))
                    .setPath(apiPath);
            if (filtersParameters.getFilters().isPresent()) {
                restParameters.addQueryParam(filtersParameters.getFilters().get());
            }
            return RestExecutorFactory.createRestExecutor(RestMethod.GET).execute(restParameters, restResponseType);
        }
        return new ArrayList<>();
    }

    protected <E> E inspectObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, String id) {
        String path = id != null ? apiPath + "/" + id : apiPath;
        RestParameters restParameters = new RestParameters(dockerWebClient.getBaseResource(swarmId))
                .setPath(path)
                        .addQueryParam(new QueryParam("insertDefaults", true));
        return RestExecutorFactory.createRestExecutor(RestMethod.GET).execute(restParameters, restResponseType);
    }

    protected <E> E inspectObject(String apiPath, String swarmId, RestResponseType<E> restResponseType) {
        return inspectObject(apiPath, swarmId, restResponseType, null);
    }

    protected <E> E createObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, RequestBodyParameter requestBodyParameter,
                                 HeaderParameters headerParameters) {
        RestParameters restParameters = createRestParameters(apiPath, swarmId, requestBodyParameter, headerParameters, null);
        return RestExecutorFactory.createRestExecutor(RestMethod.POST).execute(restParameters, restResponseType);
    }

    protected <E> E createObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, RequestBodyParameter requestBodyParameter) {
        return createObject(apiPath, swarmId, restResponseType, requestBodyParameter, null);
    }

    protected <E> E updateObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, RequestBodyParameter requestBodyParameter,
                                 QueryParameters queryParameters, HeaderParameters headerParameters) {
        RestParameters restParameters = createRestParameters(apiPath, swarmId, requestBodyParameter, headerParameters, queryParameters);
        return RestExecutorFactory.createRestExecutor(RestMethod.POST).execute(restParameters, restResponseType);
    }

    protected <E> E updateObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, RequestBodyParameter requestBodyParameter,
                               QueryParameters queryParameters) {
       return updateObject(apiPath, swarmId, restResponseType, requestBodyParameter, queryParameters, null);
    }

    protected <E> E deleteObject(String apiPath, String swarmId, RestResponseType<E> restResponseType, QueryParameters queryParameters) {
        RestParameters restParameters = createRestParameters(apiPath, swarmId, null, null, queryParameters);
        return RestExecutorFactory.createRestExecutor(RestMethod.DELETE).execute(restParameters, restResponseType);
    }

    protected <E> E deleteObject(String apiPath, String swarmId, RestResponseType<E> restResponseType) {
        return deleteObject(apiPath, swarmId, restResponseType, null);
    }

    private RestParameters createRestParameters(String apiPath, String swarmId, RequestBodyParameter requestBodyParameter,
                                                HeaderParameters headerParameters, QueryParameters queryParameters) {
        RestParameters restParameters = new RestParameters(dockerWebClient.getBaseResource(swarmId))
                .setPath(apiPath);
        if (requestBodyParameter != null && requestBodyParameter.getRequestBody() != null) {
            restParameters.setRequestBody(requestBodyParameter.getRequestBody());
        }
        if (headerParameters != null && headerParameters.getHeaders() != null) {
            restParameters.addHeaderParams(headerParameters.getHeaders());
        }
        if (queryParameters != null && queryParameters.getQueryParams() != null) {
            restParameters.addQueryParams(queryParameters.getQueryParams());
        }
        return restParameters;
    }

    public byte[] getObjectLogs(String apiPath, String swarmId, QueryParameters queryParameters) {
        RestParameters restParameters = createRestParameters(apiPath, swarmId, null, null, queryParameters);
        return RestExecutorFactory.createRestExecutor(RestMethod.GET).execute(restParameters, new RestResponseType<byte[]>(){});
    }

}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/common/AbstractApiImpl.kt
package com.swarmbit.docker.api.common

import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.parameters.FiltersParameters
import com.swarmbit.docker.api.common.parameters.HeaderParameters
import com.swarmbit.docker.api.common.parameters.QueryParameters
import com.swarmbit.docker.api.common.parameters.RequestBodyParameter
import com.swarmbit.docker.api.common.rest.executor.RestExecutorFactory
import com.swarmbit.docker.api.common.rest.model.Method
import com.swarmbit.docker.api.common.rest.model.Parameters
import com.swarmbit.docker.api.common.rest.model.QueryParam
import com.swarmbit.docker.api.common.rest.model.ResponseType

abstract class AbstractApiImpl(private val dockerWebClient: DockerWebClient) {

    protected open fun <E> listObjects(
            apiPath: String,
            swarmId: String,
            responseType: ResponseType<List<E>>,
            filtersParameters: FiltersParameters? = null
    ): List<E> {
        val restParameters: Parameters = Parameters(dockerWebClient.getBaseResource(swarmId)).setPath(apiPath)
        if (filtersParameters != null) {
            restParameters.addQueryParam(filtersParameters.filters)
        }
        return RestExecutorFactory.createRestExecutor(Method.GET).execute(restParameters, responseType) ?: emptyList()
    }

    protected open fun <E> inspectObject(
            apiPath: String,
            swarmId: String,
            responseType: ResponseType<E>,
            id: String? = null,
            queryParameters: QueryParameters? = null
    ): E? {
        val path = if (id != null) "$apiPath/$id" else apiPath
        var restParameters: Parameters = Parameters(dockerWebClient.getBaseResource(swarmId))
                .setPath(path)
        if (queryParameters != null) {
            val queryParams: List<QueryParam> = queryParameters.queryParams
            for (queryParam in queryParams) {
                restParameters = restParameters.addQueryParam(queryParam)
            }
        }
        return RestExecutorFactory.createRestExecutor(Method.GET).execute(restParameters, responseType)
    }

    protected open fun <E> createObject(
            apiPath: String,
            swarmId: String,
            responseType: ResponseType<E>,
            requestBodyParameter: RequestBodyParameter,
            headerParameters: HeaderParameters?
    ): E? {
        val parameter: Parameters = createRestParameters(apiPath, swarmId, requestBodyParameter, headerParameters, null)
        return RestExecutorFactory.createRestExecutor(Method.POST).execute(parameter, responseType)
    }

    protected open fun <E> updateObject(
            apiPath: String,
            swarmId: String,
            responseType: ResponseType<E>,
            requestBodyParameter: RequestBodyParameter,
            queryParameters: QueryParameters? = null,
            headerParameters: HeaderParameters? = null
    ): E? {
        val parameters: Parameters = createRestParameters(apiPath, swarmId, requestBodyParameter, headerParameters, queryParameters)
        return RestExecutorFactory.createRestExecutor(Method.POST).execute(parameters, responseType)
    }

    protected open fun <E> deleteObject(
            apiPath: String,
            swarmId: String,
            responseType: ResponseType<E>,
            queryParameters: QueryParameters? = null
    ): E? {
        val restParameters: Parameters = createRestParameters(apiPath, swarmId, null, null, queryParameters)
        return RestExecutorFactory.createRestExecutor(Method.DELETE).execute(restParameters, responseType)
    }

    protected open fun getObjectLogs(
            apiPath: String,
            swarmId: String,
            queryParameters: QueryParameters?
    ): ByteArray? {
        val parameters: Parameters = createRestParameters(apiPath, swarmId, null, null, queryParameters)
        return RestExecutorFactory.createRestExecutor(Method.GET).execute(parameters, object : ResponseType<ByteArray>() {})
    }

    private fun createRestParameters(
            apiPath: String,
            swarmId: String,
            requestBodyParameter: RequestBodyParameter?,
            headerParameters: HeaderParameters?,
            queryParameters: QueryParameters?
    ): Parameters {
        val parameters: Parameters = Parameters(dockerWebClient.getBaseResource(swarmId)).setPath(apiPath)
        if (requestBodyParameter?.requestBody != null) {
            parameters.setRequestBody(requestBodyParameter.requestBody)
        }
        if (headerParameters?.headers != null) {
            parameters.addHeaderParams(headerParameters.headers)
        }
        if (queryParameters?.queryParams != null) {
            parameters.addQueryParams(queryParameters.queryParams)
        }
        return parameters
    }

}